### PR TITLE
ci: run all jest tests in one operation

### PR DESCRIPTION
**Pull request by valerybugakov**
## Context

Some client unit tests [were not executed on CI](https://sourcegraph.slack.com/archives/C01N83PS4TU/p1645738079951719) because new client packages were added without updating the corresponding CI steps. To avoid this problem in the future, client unit tests are now divided into three parts:

1. `client/web`.
2. `client/browser`.
3. Everything else via the `--testPathIgnorePatterns` flag.

Also, `build-ts` and `eslint` steps are split into standalone CI steps because they do not depend on each other and can be executed independently. It makes the ESLint step ~2m faster. ⚡️

## Test plan

- Ensure that all `jest` client tests are executed on CI.
- Check out [the CI build](https://buildkite.com/sourcegraph/sourcegraph/builds/134381) that includes client steps with all changes introduced in this branch.

### CODE REVIEW

File: enterprise/dev/ci/internal/ci/operations.go [hunk 1]
```diff
@@ -226,7 +229,7 @@ func addBrowserExt(pipeline *bk.Pipeline) {
 	}
 
 	// Browser extension unit tests
-	pipeline.AddStep(":jest::chrome: Test browser extension",
+	pipeline.AddStep(":jest::chrome: Test (client/browser)",
```
Comment (valerybugakov) [inline enterprise/dev/ci/internal/ci/operations.go:232]: Titles made consistent with Go tests:

Screen Shot 2022-03-01 at 16 36 24

**Review by umpox — APPROVED** [review]
Nice 👍

File: enterprise/dev/ci/internal/ci/operations.go [hunk 2]
```diff
@@ -311,25 +314,12 @@ func clientChromaticTests(autoAcceptChanges bool) operations.Operation {
 	}
 }
 
-// Adds the shared frontend tests (shared between the web app and browser extension).
+// Adds the frontend tests (without the web app and browser extension tests).
 func frontendTests(pipeline *bk.Pipeline) {
```
Comment (bobheadxi) [inline enterprise/dev/ci/internal/ci/operations.go:318]: Should we just inline this function alongside the other `:jest: Test` steps? Or move all the `:jest: Test` sub-steps into this function?

File: dev/ci/yarn-test.sh [hunk 3]
```diff
@@ -72,4 +71,4 @@ EOF
 
 BUILDKITE_ANALYTICS_FRONTEND_UNIT_TEST_SUITE_API_KEY=$(gcloud secrets versions access latest --secret="BUILDKITE_ANALYTICS_FRONTEND_UNIT_TEST_SUITE_API_KEY" --project="sourcegraph-ci" --quiet)
 
-yarn_test
+yarn_test "$@"
```
Comment (bobheadxi) [inline dev/ci/yarn-test.sh:74]: FYI as of https://github.com/sourcegraph/sourcegraph/pull/31969 this is just the yarn command now without the need for the function (there should be a merge conflict)

**Review by bobheadxi — CHANGES_REQUESTED** [review]
See comments! Also watch out for https://github.com/sourcegraph/sourcegraph/pull/31969

File: doc/dev/background-information/ci/reference.md [hunk 4]
```diff
@@ -22,12 +22,12 @@ The default run type.
 - Pipeline for `Client` changes:
   - **Pipeline setup**: Trigger async
   - **Linters and static analysis**: Prettier, Misc linters, Yarn deduplicate lint
-  - **Client checks**: Puppeteer tests prep, Puppeteer tests chunk #1, Puppeteer tests chunk #2, Puppeteer tests chunk #3, Puppeteer tests chunk #4, Puppeteer tests chunk #5, Puppeteer tests chunk #6, Puppeteer tests chunk #7, Puppeteer tests chunk #8, Puppeteer tests chunk #9, Puppeteer tests finalize, Upload Storybook to Chromatic, Test shared client code, Test wildcard client code, Build, Enterprise build, Test, Puppeteer tests for chrome extension, Test browser extension, Test branded client code, Typescript eslint, Stylelint
+  - **Client checks**: Puppeteer tests prep, Puppeteer tests chunk #1, Puppeteer tests chunk #2, Puppeteer tests chunk #3, Puppeteer tests chunk #4, Puppeteer tests chunk #5, Puppeteer tests chunk #6, Puppeteer tests chunk #7, Puppeteer tests chunk #8, Puppeteer tests chunk #9, Puppeteer tests finalize, Upload Storybook to Chromatic, Test (all), Build, Enterprise build, Test (client/web), Puppeteer tests for chrome extension, Test (client/browser), ESLint, Build TS, Stylelint
```
Comment (limitedmage) [inline doc/dev/background-information/ci/reference.md:25]: Does `Test (all)` include `client/web` and `client/browser`?

Comment (valerybugakov) [reply doc/dev/background-information/ci/reference.md:25]: Nope, it doesn't.

File: enterprise/dev/ci/internal/ci/operations.go (hunk 2 above)
Comment (valerybugakov) [reply enterprise/dev/ci/internal/ci/operations.go:318]: Unit tests are grouped with other pipeline steps by package: e.g., all `web` package specific steps are grouped under `func addWebApp()`. I want to keep this structure in place here, and we can revisit it in a follow-up PR.

File: doc/dev/background-information/ci/reference.md (hunk 4 above)
Comment (limitedmage) [reply doc/dev/background-information/ci/reference.md:25]: Awesome, thanks for confirming! I'm not too familiar with CI code so I won't "approve" the PR, but it looks good to me on the behavior side!

File: enterprise/dev/ci/internal/ci/operations.go (hunk 2 above)
Comment (bobheadxi) [reply enterprise/dev/ci/internal/ci/operations.go:328]: Oh I see - I conflated this with the go tests, which adds everything in one operation

File: dev/ci/yarn-test.sh [hunk 5]
```diff
@@ -10,7 +10,6 @@ echo "--- generate"
 yarn gulp generate
 
 root_dir=$(pwd)
-cd "$1"
```
Comment (bobheadxi) [inline dev/ci/yarn-test.sh:13]: I think this means we can remove `root_dir` and just use relative paths instead (e.g. `JEST_JUNIT_OUTPUT_DIR="./test-reports"` on line 18)

**Review by bobheadxi — APPROVED** [review]
nice! 😁

**Pull request merged**